### PR TITLE
docs(mdCompilerProvider): add respectPreAssignBindingsEnabled docs

### LINKED ISSUE #11252: docs(mdCompilerProvider): missing from docs site
### Bug, feature request, or proposal:
Docs

### What is the expected behavior?
The docs site contains information about all important configuration details.

### What is the current behavior?
It is lacking the details about the `$mdCompilerProvider` that were added as part of https://github.com/angular/material/issues/10016 and PR https://github.com/angular/material/pull/10726. This can only be found in [the code comments](https://github.com/angular/material/blob/master/src/core/services/compiler/compiler.js#L84-L167) or [CHANGELOG](https://github.com/angular/material/blob/master/CHANGELOG.md#mdcompiler) currently. It needs to be documented on the main docs site as well.

### Which versions of AngularJS, Material, OS, and browsers are affected?
AngularJS 1.6.x
AngularJS Material 1.1.5+

### Is there anything else we should know? Stack Traces, Screenshots, etc.
N/A

### PR DESCRIPTION
**Pull request by Splaktar**
## PR Checklist
Please check that your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/angular/material/blob/master/.github/CONTRIBUTING.md#-commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
The docs site is missing the details about `$mdCompilerProvider.respectPreAssignBindingsEnabled`. The feature was added as part of https://github.com/angular/material/issues/10016 and PR https://github.com/angular/material/pull/10726. These docs can only be found in [the code comments](https://github.com/angular/material/blob/master/src/core/services/compiler/compiler.js#L84-L167) or [CHANGELOG](https://github.com/angular/material/blob/master/CHANGELOG.md#mdcompiler) currently. It needs to be documented on the main docs site as well.

Issue Number: 
Fixes #11252.

## What is the new behavior?
`$mdCompilerProvider.respectPreAssignBindingsEnabled` is documented on the new `$mdCompilerProvider` docs page under Services.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
N/A

### CODE REVIEW

**Comment by Splaktar** [comment]
@gkalpak can you please take a look at these docs changes to verify that they are still accurate with AngularJS 1.6 and the plans for AngularJS 1.7?

File: src/core/services/compiler/compiler.js [hunk 1]
```diff
@@ -87,39 +103,37 @@ function MdCompilerProvider($compileProvider) {
...
-   * @name  $mdCompilerProvider#respectPreAssignBindingsEnabled
+   * @ngdoc method
+   * @name $mdCompilerProvider#respectPreAssignBindingsEnabled
    *
-   * @param {boolean=} respected update the respectPreAssignBindingsEnabled state if provided, otherwise just return
-   * the current Material preAssignBindingsEnabled state
-   * @returns {*} current value if used as getter or itself (chaining) if used as setter
+   * @param {boolean=} respected update the `respectPreAssignBindingsEnabled` state if provided, otherwise just return
+   * the current Material `respectPreAssignBindingsEnabled` state.
+   * @returns {boolean|function} current value if used as getter or itself (chaining) if used as setter
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: Why `function`? It returns the `MdCompilerProvider` instance if used as setter.

File: src/core/services/compiler/compiler.js [hunk 2]
```diff
@@ -10,74 +10,90 @@ angular
 
 /**
  * @ngdoc service
- * @name $mdCompiler
+ * @name $mdCompilerProvider
  * @module material.core.compiler
  * @description
- * The $mdCompiler service is an abstraction of AngularJS's compiler, that allows developers
- * to easily compile an element with options like in a Directive Definition Object.
+ * The `$mdCompiler` is able to respect the AngularJS `$compiler.preAssignBindingsEnabled` state when using
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: `$compiler.preAssignBindingsEnabled` --> $compileProvider.preAssignBindingsEnabled` (here and elsewhere).

File: src/core/services/compiler/compiler.js [hunk 3]
```diff
@@ -10,74 +10,90 @@ angular
...
+ * (i.e. dialog, panel, toast, bottomsheet) respect the AngularJS `$compile.preAssignBindingsEnabled` flag,
+ * call the AngularJS Material method: `$mdCompilerProvider.respectPreAssignBindingsEnabled(boolean)`.
+ *
+ * This AngularJS Material *flag* doesn't affect directives/components created via regular AngularJS methods.
+ * These constitute the majority of AngularJS Material and user-created components.
+ *
+ * Only dynamic construction of elements such as Dialogs, Panels, Toasts, BottomSheets, etc. may be affected.
+ * Invoking `$mdCompilerProvider.respectPreAssignBindingsEnabled(true)` will make **bindings** in
+ * AngularJS Material custom components like `$mdDialog`, `$mdPanel`, `$mdToast`, or `$mdBottomSheet`
+ * only available in the controller constructors.
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: I believe this is wrong. (As mentioned below, invoking `$mdCompilerProvider.respectPreAssignBindingsEnabled(true)` can result in different behaviors, depending on `$compileProvider.preAssignBindingsEnabled()` and AngularJS version.)

File: src/core/services/compiler/compiler.js [hunk 4]
```diff
@@ -10,74 +10,90 @@ angular
...
- *     templateUrl: 'modal.html',
- *     controller: 'ModalCtrl',
- *     locals: {
- *       modal: myModalInstance;
- *     }
- *   }).then(function (compileData) {
- *     compileData.element; // Compiled DOM element
- *     compileData.link(myScope); // Instantiate controller and link element to scope.
+ *   app.config(function($mdCompilerProvider) {
+ *    $mdCompilerProvider.respectPreAssignBindingsEnabled(true);
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: Insufficient indentation 😇

File: src/core/services/compiler/compiler.js [hunk 5]
```diff
@@ -10,74 +10,90 @@ angular
...
- *   // The compiler doesn't need to recompile the element each time.
- *   var myElement = $compile('<span>Test</span>')(myScope);
- *
- *   $mdCompiler.compile({
- *     contentElement: myElement
- *   }).then(function (compileData) {
- *     compileData.element // Content Element (same as above)
- *     compileData.link // This does nothing when using a contentElement.
+ *   app.config(function($mdCompilerProvider) {
+ *    $mdCompilerProvider.respectPreAssignBindingsEnabled(false);
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: Insufficient indentation. Also, this does not guarrantee that bindings are pre-assigned.

File: src/core/services/compiler/compiler.js [hunk 6]
```diff
@@ -10,74 +10,90 @@ angular
...
+ * }
  *
- *   $mdCompiler.compile({
- *     contentElement: myElement
- *   }).then(function (compileData) {
- *     compileData.element // Content Element (same as above)
- *     compileData.link // This does nothing when using a contentElement.
- *   });
+ * function MyController() {
+ *   // Locals from Angular Material are set yet. e.g myVar is true.
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: _are set yet_ --> _are set already_ (?)

File: src/core/services/compiler/compiler.js (hunk 5 above)
Comment (Splaktar) [reply src/core/services/compiler/compiler.js]: I believe that when `$mdCompilerProvider.respectPreAssignBindingsEnabled(false)` is called, it should guarantee that the bindings of AngularJS Material interim components are pre-assigned in AngularJS 1.5 and 1.6. Is this not correct? The description on line 119-120 seems pretty clear that the disabled (`false`) option doesn't depend on AngularJS versions or settings.

File: src/core/services/compiler/compiler.js (hunk 3 above)
Comment (Splaktar) [reply src/core/services/compiler/compiler.js]: I agree, I've cleaned this up.

File: src/core/services/compiler/compiler.js [hunk 7]
```diff
@@ -10,74 +10,91 @@ angular
...
  * @ngdoc service
- * @name $mdCompiler
+ * @name $mdCompilerProvider
  * @module material.core.compiler
  * @description
- * The $mdCompiler service is an abstraction of AngularJS's compiler, that allows developers
- * to easily compile an element with options like in a Directive Definition Object.
+ * The `$mdCompiler` is able to respect the AngularJS `$compileProvider.preAssignBindingsEnabled` state when using
+ * AngularJS versions between 1.5.10 and 1.7.0.
+ * See the [AngularJS documentation for `$compile.preAssignBindingsEnabled`
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: `$compile` --> `$compileProvider`

File: src/core/services/compiler/compiler.js [hunk 8]
```diff
@@ -10,74 +10,91 @@ angular
...
+ * ](https://code.angularjs.org/1.6.10/docs/api/ng/provider/$compileProvider#preAssignBindingsEnabled)
+ * for more information.
  *
- * > The compiler powers a lot of components inside of AngularJS Material.
- * > Like the `$mdPanel` or `$mdDialog`.
+ * To enable/disable whether the controllers of dynamic AngularJS Material components
+ * (i.e. dialog, panel, toast, bottomsheet) respect the AngularJS `$compile.preAssignBindingsEnabled` flag,
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: `$compile` --> `$compileProvider`

File: src/core/services/compiler/compiler.js [hunk 9]
```diff
@@ -87,39 +104,37 @@ function MdCompilerProvider($compileProvider) {
...
    * @description
-   * Call this method to enable/disable whether Material-specific (dialogs/toasts) controllers respect the AngularJS
-   * `$compile.preAssignBindingsEnabled` flag. Note that this doesn't affect directives/components created via
-   * regular AngularJS methods which constitute most Material & user-created components.
-   *
-   * @see [AngularJS documentation for `$compile.preAssignBindingsEnabled`
-   * ](https://code.angularjs.org/1.6.4/docs/api/ng/provider/$compileProvider#preAssignBindingsEnabled)
-   * for more information.
+   * Call this method to enable/disable whether Material-specific (dialog/panel/toast/bottomsheet) controllers respect the
+   * AngularJS `$compile.preAssignBindingsEnabled` flag. Note that this doesn't affect directives/components created
```
Comment (gkalpak) [inline src/core/services/compiler/compiler.js]: `$compile` --> `$compileProvider`

File: src/core/services/compiler/compiler.js [hunk 10]
```diff
@@ -10,74 +10,91 @@ angular
...
 /**
  * @ngdoc service
- * @name $mdCompiler
+ * @name $mdCompilerProvider
  * @module material.core.compiler
  * @description
- * The $mdCompiler service is an abstraction of AngularJS's compiler, that allows developers
- * to easily compile an element with options like in a Directive Definition Object.
+ * The `$mdCompiler` is able to respect the AngularJS `$compileProvider.preAssignBindingsEnabled` state when using
+ * AngularJS versions between 1.5.10 and 1.7.0.
```
Comment (mgol) [inline src/core/services/compiler/compiler.js]: It sounds like it respects it for 1.7.0 as well. How about:
```js
 * The `$mdCompiler` is able to respect the AngularJS `$compileProvider.preAssignBindingsEnabled` state when using
 * AngularJS versions from 1.5.10 until (but not including) 1.7.0.
```
?

File: src/core/services/compiler/compiler.js [hunk 11]
```diff
@@ -10,74 +10,91 @@ angular
...
- * > Like the `$mdPanel` or `$mdDialog`.
+ * To enable/disable whether the controllers of dynamic AngularJS Material components
+ * (i.e. dialog, panel, toast, bottomsheet) respect the AngularJS `$compile.preAssignBindingsEnabled` flag,
+ * call the AngularJS Material method: `$mdCompilerProvider.respectPreAssignBindingsEnabled(boolean)`.
+ *
+ * This AngularJS Material *flag* doesn't affect directives/components created via regular AngularJS methods.
+ * These constitute the majority of AngularJS Material and user-created components.
+ * Only dynamic construction of elements such as Dialogs, Panels, Toasts, BottomSheets, etc. may be affected.
+ * Invoking `$mdCompilerProvider.respectPreAssignBindingsEnabled(true)` will effect **bindings** in
+ * AngularJS Material custom components like `$mdDialog`, `$mdPanel`, `$mdToast`, or `$mdBottomSheet`.
```
Comment (mgol) [inline src/core/services/compiler/compiler.js]: `$mdDialog` etc. are services, not components. They don't even show components but usually templates with controllers (it'd be nice to add support for components but that's a separate story). Maybe:

> ...\*\*bindings\*\* in controllers created by AngularJS Material's services like `$mdDialog`, `$mdPanel`, `$mdToast`, or `$mdBottomSheet`.

?

File: src/core/services/compiler/compiler.js [hunk 12]
```diff
@@ -10,74 +10,91 @@ angular
...
- *   }).then(function (compileData) {
- *     compileData.element; // Compiled DOM element
- *     compileData.link(myScope); // Instantiate controller and link element to scope.
+ *   app.config(function($mdCompilerProvider) {
+ *     $mdCompilerProvider.respectPreAssignBindingsEnabled(true);
  *   });
  * </hljs>
  *
- * Example with a content element
+ * Assign Bindings Before the Constructor
```
Comment (mgol) [inline src/core/services/compiler/compiler.js]: This is only true for AngularJS <1.6.0 or AngularJS <1.7.0 if `$compileProvider.preAssignBindingsEnabled(true)` is called. Shouldn't it just say the reverse of the upper example's "Respect the AngularJS Compiler Setting"?

File: src/core/services/compiler/compiler.js (hunk 10 above)
Comment (Splaktar) [reply src/core/services/compiler/compiler.js]: `$compileProvider.preAssignBindingsEnabled` is removed from AngularJS in 1.7.0. Maybe that's your point? I'll try to refine it.

File: src/core/services/compiler/compiler.js (hunk 7 above)
Comment (Splaktar) [reply src/core/services/compiler/compiler.js]: Doh! Not sure how I missed those, thanks.

**Comment by mgol** [comment]
@andrewseguin @Splaktar It seems none of my comments were addressed in the merged commit; do you plan followups?

**Pull request merged**